ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/PartitionBy.java

```diff
@@ -17,27 +17,43 @@
 
 import static java.util.Objects.requireNonNull;
 
+import com.google.common.collect.ImmutableList;
 import com.google.errorprone.annotations.Immutable;
 import io.confluent.ksql.execution.expression.tree.Expression;
 import io.confluent.ksql.parser.NodeLocation;
+import io.confluent.ksql.util.KsqlException;
+import java.util.HashSet;
+import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 
 @Immutable
 public class PartitionBy extends AstNode {
 
-  private final Expression expression;
+  private final List<Expression> partitionByExpressions;
 
   public PartitionBy(
       final Optional<NodeLocation> location,
-      final Expression partitionBy
+      final List<Expression> partitionByExpressions
   ) {
     super(location);
-    this.expression = requireNonNull(partitionBy, "partitionBy");
+    this.partitionByExpressions = ImmutableList
+        .copyOf(requireNonNull(partitionByExpressions, "partitionByExpressions"));
+
+    if (partitionByExpressions.isEmpty()) {
+      throw new KsqlException("PARTITION BY requires at least one expression");
+    }
+
+    final HashSet<Object> partitionBys = new HashSet<>(partitionByExpressions.size());
+    partitionByExpressions.forEach(exp -> {
+      if (!partitionBys.add(exp)) {
+        throw new KsqlException("Duplicate PARTITION BY expression: " + exp);
```

Comment:
what's wrong with a duplicate partition by expression? I don't see any reason why a user might want it, but I don't see why not either (e.g. maybe their output data expects `userId, userSpecialId` in the key and this stream always has the same value for both)

I get that the key name conflicts might be a little weird, so we can do this in a follow-up PR, but I don't think we should prohibit it

Reply:
This is consistent with what we do for multi-column GROUP BY (which has been around for a long time): https://github.com/confluentinc/ksql/blob/e7f1f4786e07463128d1ea73a010757406fc180e/ksqldb-parser/src/main/java/io/confluent/ksql/parser/tree/GroupBy.java#L49-L53

I believe the reason indeed has to do with naming.

Reply:
sounds good, we can keep it like that for now